### DISCUSSION TITLE
How to either control the node insertion or swap nodes?

### DISCUSSION THREAD

**Discussion by steweg** [CESNET/libyang#2233] [Q&A]
Hi,

I am trying to perform replacement of particular subtree within the data tree with a brand new one while I wish to preserve node ordering regardless of associated schema. As this is not possible natively (there is no `lyd_change_inner()` nor `lyd_swap_node()` API), my workaround consist of the following steps:
1. Create new subtree
2. Insert new subtree
3. Unlink and delete old subtree

The problematic is the 2. step. It requires to have full control of node positioning, which is not allowed:
- In case when there is no parent (== top level nodes)
  - The `lyd_insert_after()` or `lyd_insert_before()` is acceptable only when associated schema is list/leaf-list with `ordered-by user;`
  - In other cases the order is controlled by libyang (e.g. within `lyd_insert_sibling()`)
- In case when there is a parent, then the behavior is controlled completely by libyang itself during creation of nodes (e.g. within `lyd_new_term()`)

Alternative approach that I also consider was like this:
- iterate over new subtree children and change parent reference to old subtree root
- iterate over old subtree children and change parent reference to new subtree root
- swap the child references of new and old subtree
- unlink and delete new subtree
I am not sure if this will work or not correctly as there is also the node hashing part, for which I have no clue how to should work, but I guess something needs to be done and as this is not able to be done via libyang public APIs.... So it would mean I would need to create a libyang patch first...

Another alternative was to swap the references of `prev` & `next` of old and new subtrees, and also change the `child` reference of parent in case old subtree is first child. But again the question about the hashing pops up.

I was also considering creating a diff between those two subtrees and then apply diff or original subtree, but it seems to be relatively expensive operation as it involves creation of third subtree...

So do you have some better ideas how to achieve the needed result? Thanks

**Comment by michalvasko** [comment]
> I am trying to perform replacement of particular subtree within the data tree with a brand new one while I wish to preserve node ordering regardless of associated schema.

You cannot do this, *libyang* enforces and then always expects nodes to follow their schema ordering and for system-ordered (leaf-)lists, even the relative order of all the instances is strict. So my first question would be why do you want to use some other node ordering.

**Comment by steweg** [comment]
I am ok with default libyang ordering. No problems with that. I am just trying to achieve swaping of particular subtree with a new one (the root nodes of old and new subtrees are always using the same schema)

**Reply by michalvasko** [reply]
Then I am not sure what the problem is, this should work.
```C
struct lyd_node *old; // old subtree root
struct lyd_node *new; // new subtree root
struct lyd_node *parent;

parent = lyd_parent(old);
if (parent) {
  lyd_unlink_tree(old);
  lyd_insert_child(parent, new);
} else {
  lyd_insert_sibling(new, old, NULL);
  lyd_unlink_tree(old);
}
```
Actually, you do not even need to use the `parent` and just do the `else` code but it depends on your exact use-case.

**Comment by steweg** [comment]
Thanks for the hint about the libyang native ordering mechanism. I though that it is always the putting it as last. Now it make sense. I was also not aware about `lyd_insert_XXX()` will automatically unlink. I somehow missed this info in the documentation (my mistake).

So all that is left is that I just need to handle a special case when I am replacing subtree, within `list` or `leaf-list` with `ordered-by user;`.

Thanks for the hints
